File: 216-combination_sum_iii.py
# time: factorial
# space: O(k), keep track of current combo + recursive stack, does not account for collecting the result list
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def backtrack(start, combo):
            if len(combo) == k and sum(combo) == n:
                combos.append(combo[:])
            elif len(combo) > k or sum(combo) > n:
                return

            for num in range(start, 10):
                combo.append(num)
                backtrack(num + 1, combo)
                combo.pop()

        combos = []
        backtrack(1, [])
        return combos

File: 216-combination_sum_iii.cs
public class Solution {
    public IList<IList<int>> CombinationSum3(int k, int n) {
        List<IList<int>> combos = new();
        backtrack(1, n, new());
        return combos;

        void backtrack(int start, int remain, Stack<int> combo) {
            if (combo.Count == k && remain == 0) {
                combos.Add(new List<int>(combo));
                return;
            }
            if (combo.Count > k || remain < 0) {
                return;
            }

            for (int num = start; num < 10; num++) {
                combo.Push(num);
                backtrack(num + 1, remain - num, combo);
                combo.Pop();
            }
        }
    }
}
